File: DesignPatternsExercises/ruby/observer_design_pattern/solution/breaking_news_feed.rb
class BreakingNewsFeed
  # attr_reader __________(1)__________
  attr_reader :breaking_news

  def initialize
    @observers = []
  end

  def breaking_news=(breaking_news)
    @breaking_news = breaking_news
    # __________(2)__________
    notify_subscribers
  end

  def subscribe(observer)
    @observers << observer
  end

  private

  def notify_subscribers
    # __________(3)__________.each(&:send_notification)
    @observers.each(&:send_notification)
  end
end

File: DesignPatternsExercises/java/ObserverDesignPattern/Solution/BreakingNewsFeed.java
import java.util.ArrayList;
import java.util.List;

public class BreakingNewsFeed {
    String breakingNews;
    // List<__________(1)__________> observers = new ArrayList<>();
    List<Observer> observers = new ArrayList<>();

    public String getBreakingNews() {
        return breakingNews;
    }

    public void setBreakingNews(String breakingNews) {
        this.breakingNews = breakingNews;
        // __________(2)__________;
        notifySubscribers();
    }

    public void subscribe(Observer observer) {
        observers.add(observer);
    }

    private void notifySubscribers() {
        for (Observer observer : observers)
            // observer.__________(3)__________();
            observer.sendNotification();
    }
}
